src/main/python/hydra/zmqtest/runtest.py

```diff
@@ -26,42 +28,111 @@ def __init__(self, argv):
+        self.add_appid(self.zstsub)
+        #self.total_sub_apps = 200
+        self.total_sub_apps = zmq_config.total_apps
         self.start_init()
+        self.run_test()
+
+    def run_test(self):
+        # Launch zmq pub
+        self.launch_zmq_pub()
+
+        # Launch zmq sub up to self.total_sub_apps
+        self.launch_zmq_sub()
+
+        # Signal PUB to start sending all messages, blocks until PUB notifies
+        l.info("Waiting for 10 secs before signalling PUB to send messages ..")
+        time.sleep(10)
+        self.signal_pub_send_msgs()
+
+        #  Fetch all client SUB data
+        self.fetch_all_sub_clients_data()
+
+        #  Delete all launched apps
+        self.delete_all_launched_apps()
+        l.info("Successfully finished gathering all data")
+
+        l.info("================================================")
+        for client, info in self.all_sub_clients_info.items():
+            l.info("Client: %s", client)
+            l.info("Info: %s", info)
+        l.info("================================================")
+        l.info("Please press ctrl+z  and then \"kill %1\"  twice.. sorry, will figure this out soon")
+        return
+
+    def launch_zmq_pub(self):
         l.info("Launching the pub app")
-        self.create_hydra_app(name=zstpub, app_path='hydra.zmqtest.zmqtests.zmq_pub', app_args='1555',
+        self.create_hydra_app(name=self.zstpub, app_path='hydra.zmqtest.zmq_pub.run',
+                              app_args='0.0001', # msg delay, msg count
                               cpus=0.01, mem=32,
+                              ports=[0],
                               constraints=[self.app_constraints(field='hostname', operator='UNIQUE')])
+        self.wait_app_ready(self.zstpub, 1)
+        # Find launched pub server's ip, rep PORT
+        self.pub_ip = self.find_ip_uniqueapp(self.zstpub)
+        tasks = self.get_app_tasks(self.zstpub)
+        assert(len(tasks) == 1)
+        assert(len(tasks[0].ports) == 1)
+        self.pub_rep_taskport = str(tasks[0].ports[0])
+        l.info("[zmq_pub] ZMQ pub server running at [%s]", self.pub_ip)
+        l.info("[zmq_pub] ZMQ REP server running at [%s:%s]", self.pub_ip, self.pub_rep_taskport)
+        time.sleep(2)
+
+    def launch_zmq_sub(self):
+        l.info("Launching the sub app")
+        self.create_hydra_app(name=self.zstsub, app_path='hydra.zmqtest.zmq_sub.run',
+                              app_args='%s 15556' % (self.pub_ip), # pub_ip, pub_port
+                              cpus=0.01, mem=32,
+                              ports=[0])
+        self.wait_app_ready(self.zstsub, 1)
+
+        # scale
+        l.info("Scaling sub app to [%d]", self.total_sub_apps)
+        self.scale_app(self.zstsub, self.total_sub_apps)
+        self.wait_app_ready(self.zstsub, self.total_sub_apps)
+
+        # Extract ip, rep server port for data querying
+        self.sub_app_ip_rep_port_map = {} # key = port, value=ip
```

Comment:
port can't be the key, port will be reused across nodes, it should be the taksid